### PR TITLE
feat: support sft for embodiment

### PR DESCRIPTION
**Pull request by XuS1994**
feat: Support SFT training with ManiSkill Open-PI dataset

Description

This PR adds full support for Supervised Fine-Tuning (SFT) using the ManiSkill Open-PI human-demonstration dataset.

Key points:

New data-loader ManiSkillOPDataset under rlinf/data/ that
ingests Open-PI jsonl/npz files
tokenises observation–action pairs into RLInf’s dialogue format
supports on-the-fly filtering, shuffling and bucketing for efficient GPU utilisation.

Training pipeline updates
scripts/train_sft.py now accepts --dataset mani_openpi and related CLI flags (obs modality, clip length, max tokens, etc.).
Added SFTTrainer.build_dataloader() logic to plug in any dataset that inherits the new BaseSFTDataset interface.

Docs & examples
New example config examples/coding_online_rl/sft_mani_openpi.yaml.
English & Chinese docs: “docs/source-en/sft_openpi.rst” & “docs/source-zh/sft_openpi.rst” with step-by-step tutorial.

Motivation and Context
Supervised Fine-Tuning on large-scale human demonstration datasets is a crucial stepping-stone before RLHF or RLAIF. ManiSkill Open-PI provides 4.5 M high-fidelity manipulation trajectories covering 20+ tasks. Native support lets researchers reproduce Open-PI baselines, benchmark model quality and bootstrap downstream RL phases with stronger priors.

How has this been tested?

E2E tests
tests/e2e_tests/sft_mani_openpi.yaml trains a tiny (125 M) model for 50 steps on CI; asserts loss decreases.

Types of changes

[ ] Bug fix (non-breaking change which fixes an issue)
[x] New feature (non-breaking change which adds functionality)
[ ] Documentation update (Document-only update)
[ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
Checklist

[x] My code follows the code style of this project.
[x] My change requires a change to the documentation.
[x] I have updated the documentation accordingly.
[x] I have added tests to cover my changes.
[x] All new and existing tests passed.

### CODE REVIEW

File: docs/source-en/rst_source/examples/sft.rst [hunk 1]
```diff
@@ -0,0 +1,108 @@
+
+Supervised Fine-Tuning (SFT)
+============================
+
+.. |huggingface| image:: /_static/svg/hf-logo.svg
+   :width: 16px
+   :height: 16px
+   :class: inline-icon
+
+This document explains how to run **full-parameter supervised fine-tuning (SFT)** in the RLinf framework. SFT is typically the first stage before RLHF / RLAIF: the model first imitates high-quality demonstrations, then reinforcement learning continues optimizing from that prior.
```
Comment (zoeyuchao) [inline docs/source-en/rst_source/examples/sft.rst]: 我们是不是lora微调也支持？
另外，readme里面需要把SFT打上对号/

Comment (XuS1994) [reply docs/source-en/rst_source/examples/sft.rst]: Yes, I added basic lora training step in doc

File: rlinf/models/embodiment/openpi/__init__.py [hunk 2]
```diff
@@ -143,6 +158,20 @@
...
+    TrainConfig(
+        name="pi0_custom",
+        model=pi0_config.Pi0Config(),
+        data=CustomDataConfig(
+            repo_id="physical-intelligence/custom_dataset",
+            base_config=DataConfig(
+                prompt_from_task=True
+            ),  # we need language instruction
+            assets=AssetsConfig(assets_dir="checkpoints/torch/pi0_base/assets"),
+            raw_action_is_delta=True,  # True for delta action, False for abs_action
```
Comment (UltronAI) [inline rlinf/models/embodiment/openpi/__init__.py]: 这里命名与extra_delta_transform不一致，另外这两个参数是franka data config所特有的，不是共用的参数，最好也在文档里说明一下

File: rlinf/runners/sft_runner.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+from rlinf.workers.actor.fsdp_sft_worker import SFTactor
+
+
+class SFTRunner:
+    def __init__(
+        self,
+        cfg: DictConfig,
+        actor: SFTactor,
+        critic: Optional[object] = None,
+        reward: Optional[object] = None,
```
Comment (UltronAI) [inline rlinf/runners/sft_runner.py]: 需要在sft里保留吗？

File: rlinf/runners/sft_runner.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        self.global_step = int(resume_dir.split("global_step_")[-1])
+
+    def _sync_weights(self) -> None:
+        rollout_handle: Handle = self.rollout.sync_model_from_actor()
+        actor_handle: Handle = self.actor.sync_model_to_rollout()
+        actor_handle.wait()
+        rollout_handle.wait()
+
+    def evaluate(self) -> dict[str, float]:
+        env_handle: Handle = self.env.evaluate(
```
Comment (andylin-hao) [inline rlinf/runners/sft_runner.py:75]: Need to remove this function. The runner doesn't have env & rollout

File: examples/sft/config/custom_sft_openpi.yaml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  data_path: "/path/to/custom-data"
+
+algorithm:
+  adv_type: gae
+
+actor:
+  group_name: "ActorGroup"
+  training_backend: "fsdp"
+  micro_batch_size: 1
+  global_batch_size: 16
```
Comment (guozhen1997) [inline examples/sft/config/custom_sft_openpi.yaml:41]: `global_batch_size` is set but not used properly. It should be applied when building the OpenPI dataloader. Currently, in `openpi_data_loader.create_data_loader` if `config.batch_size` is not explicitly set, the default value `batch_size: int = 32` is used, which results in `global_batch_size=32` instead of 16 as specified in the original yaml.

File: rlinf/workers/sft/fsdp_sft_worker.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        if SupportedModel(self.cfg.actor.model.model_type) in [SupportedModel.OPENPI]:
+            import openpi.training.data_loader as openpi_data_loader
+
+            from rlinf.models.embodiment.openpi import get_openpi_config
+
+            config = get_openpi_config(
+                self.cfg.actor.model.openpi.config_name,
+                model_path=self.cfg.actor.model.model_path,
+            )
+            data_loader = openpi_data_loader.create_data_loader(
```
Comment (guozhen1997) [inline rlinf/workers/sft/fsdp_sft_worker.py:78]: corresponding to the comment for `global_batch_size` setting in yaml

File: rlinf/workers/actor/fsdp_sft_worker.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+)
+from rlinf.utils.placement import (
+    HybridComponentPlacement,
+)
+from rlinf.utils.utils import (
+    clear_memory,
+)
+
+
+class SFTactor(FSDPModelManager, Worker):
```
Comment (guozhen1997) [inline rlinf/workers/actor/fsdp_sft_worker.py]: Suggest rename as `FSDPSftWorker`, which will be used training both actor and critic models.

File: rlinf/workers/sft/fsdp_sft_worker.py
Comment (guozhen1997) [inline rlinf/workers/sft/fsdp_sft_worker.py:1]: Suggest move this file into `workers/sft/` folder.

File: rlinf/workers/actor/fsdp_sft_worker.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    def __init__(self, cfg: DictConfig):
+        Worker.__init__(self)
+        super().__init__(cfg.actor, self._world_size, self._rank)
+
+        self.cfg = cfg
+        torch.cuda.set_device(int(os.environ["LOCAL_RANK"]))
+        self.device = torch.cuda.current_device()
+        self.device_mesh = init_device_mesh(
```
Comment (guozhen1997) [inline rlinf/workers/actor/fsdp_sft_worker.py]: It is not necessary to `init_device_mesh` here, since the fsdp device mesh has been initialized in L45 `super().__init__`.

File: rlinf/workers/sft/fsdp_sft_worker.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+            for _ in range(self.gradient_accumulation):
+                observation, actions = next(self.data_iter)
+
+                observation = jax.tree.map(
+                    lambda x: torch.as_tensor(x, device=self.device)
+                    .contiguous()
+                    .clone(),
+                    observation,
+                )  # noqa: PLW2901
```
Comment (UltronAI) [inline rlinf/workers/sft/fsdp_sft_worker.py]: 可以换成pure torch的实现

File: examples/sft/config/custom_sft_openpi.yaml (hunk 5 above)
Comment (XuS1994) [reply examples/sft/config/custom_sft_openpi.yaml:41]: I want to train global_batch_size data in each loop, so loop times are `global_batch_size // micro_batch_size // world_size`

 I load data once in each loop, so `config.batch_size` should be set to `micro_batch_size`.
 
 Is that right?

File: rlinf/workers/sft/fsdp_sft_worker.py (hunk 6 above)
Comment (XuS1994) [reply rlinf/workers/sft/fsdp_sft_worker.py:78]: should be in my opinion.
```
            config = get_openpi_config(
                self.cfg.actor.model.openpi.config_name,
                model_path=self.cfg.actor.model.model_path,
                batch_size=self.cfg.actor.model.batch_size,
            )
```

File: rlinf/workers/actor/fsdp_sft_worker.py (hunk 7 above)
Comment (XuS1994) [reply rlinf/workers/actor/fsdp_sft_worker.py]: OK

File: rlinf/workers/sft/fsdp_sft_worker.py
Comment (XuS1994) [reply rlinf/workers/sft/fsdp_sft_worker.py:1]: OK

File: rlinf/workers/actor/fsdp_sft_worker.py (hunk 8 above)
Comment (XuS1994) [reply rlinf/workers/actor/fsdp_sft_worker.py]: OK, but FSDPActor init device in __init__ at `rlinf/workers/actor/fsdp_actor_worker.py:89`

File: rlinf/models/embodiment/openpi/__init__.py (hunk 2 above)
Comment (XuS1994) [reply rlinf/models/embodiment/openpi/__init__.py]: align to `extra_delta_transform`.

and add comments for user extra config.

File: rlinf/runners/sft_runner.py (hunk 3 above)
Comment (XuS1994) [reply rlinf/runners/sft_runner.py]: Do not need, delete them.

File: rlinf/workers/sft/fsdp_sft_worker.py (hunk 9 above)
Comment (XuS1994) [reply rlinf/workers/sft/fsdp_sft_worker.py]: jax.tree.map also used in `openpi_action_model.py`, we can change them together if need.

Comment (UltronAI) [reply rlinf/workers/sft/fsdp_sft_worker.py]: > jax.tree.map also used in `openpi_action_model.py`, we can change them together if need.

It’s up to you. I would prefer to remove the jax dependency from this repo, since it appears to be used only for jax.tree.map (unless I’m missing something), which can be easily replaced with a pure torch implementation. Just for a cleaner style.

File: rlinf/models/embodiment/openpi/dataconfig/franka_dataconfig.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        self, assets_dirs: pathlib.Path, model_config: _model.BaseModelConfig
+    ) -> DataConfig:
+        # The repack transform is *only* applied to the data coming from the dataset,
+        # and *not* during inference. We can use it to make inputs from the dataset look
+        # as close as possible to those coming from the inference environment (e.g. match the keys).
+        # Below, we match the keys in the dataset (which we defined in the data conversion script) to
+        # the keys we use in our inference pipeline (defined in the inference script for libero).
+        # For your own dataset, first figure out what keys your environment passes to the policy server
+        # and then modify the mappings below so your dataset's keys get matched to those target keys.
+        # The repack transform simply remaps key names here.
```
Comment (chenkang455) [inline rlinf/models/embodiment/openpi/dataconfig/franka_dataconfig.py]: Can we delete these comments like https://github.com/RLinf/RLinf/blob/main/rlinf/models/embodiment/openpi/policies/calvin_policy.py ?  We left them on the libero_policy and keep other policies as simple as possible.

File: examples/sft/config/libero_sft_openpi.yaml [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+      config_name: "pi0_libero"
+      detach_critic_input: True
+
+  optim:
+    lr: 7.91e-6
+    value_lr: 1.55e-4
+    adam_beta1: 0.9
+    adam_beta2: 0.95
+    adam_eps: 1.0e-05
+    clip_grad: 1.0
```
Comment (chenkang455) [inline examples/sft/config/libero_sft_openpi.yaml:60]: They are not sft related lr. Can we delete them? Besides, if i want to adjust the training params in sft like lr or bs, how can i adjust them?

**Review by zoeyuchao — APPROVED** [review]

File: examples/sft/config/libero_sft_openpi.yaml (hunk 11 above)
Comment (XuS1994) [reply examples/sft/config/libero_sft_openpi.yaml:60]: We use same api to `get_model` and `build_optimizer`

So, RL and SFT use same param.

File: rlinf/models/embodiment/openpi/dataconfig/franka_dataconfig.py (hunk 10 above)
Comment (XuS1994) [reply rlinf/models/embodiment/openpi/dataconfig/franka_dataconfig.py]: OK

File: examples/sft/config/libero_sft_openpi.yaml [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+algorithm:
+  adv_type: gae
+
+actor:
+  group_name: "ActorGroup"
+  training_backend: "fsdp"
+  micro_batch_size: 1
+  global_batch_size: 1024
+  seed: 0
+  log_interval: 1
```
Comment (guozhen1997) [inline examples/sft/config/libero_sft_openpi.yaml]: remove this config

Comment (XuS1994) [reply examples/sft/config/libero_sft_openpi.yaml]: done

File: examples/sft/run_embodiment_sft.sh [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+export EMBODIED_PATH="$( cd "$(dirname "${BASH_SOURCE[0]}" )" && pwd )"
+export REPO_PATH=$(dirname $(dirname "$EMBODIED_PATH"))
+export SRC_FILE="${EMBODIED_PATH}/train_embodied_sft.py"
+
+export MUJOCO_GL="egl"
+export PYOPENGL_PLATFORM="egl"
+
+unset HF_LEROBOT_HOME
+# export HF_LEROBOT_HOME="/mnt/mnt/public/xusi/openpi-main/data"
```
Comment (guozhen1997) [inline examples/sft/run_embodiment_sft.sh]: remove L11 and L14

File: rlinf/models/embodiment/openpi/policies/maniskill_policy.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+        if "actions" in data:
+            inputs["actions"] = data["actions"]
+
+        # Pass the prompt (aka language instruction) to the model.
+        # Keep this for your own dataset (but modify the key if the instruction is not
+        # stored in "prompt"; the output dict always needs to have the key "prompt").
+        if "prompt" in data:
+            inputs["prompt"] = data["prompt"]
+
+        # print(f"inputs={inputs.keys()}")
```
Comment (guozhen1997) [inline rlinf/models/embodiment/openpi/policies/maniskill_policy.py]: remove L108-111

File: rlinf/workers/sft/fsdp_sft_worker.py [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+            self.gradient_accumulation = (
+                self.cfg.actor.global_batch_size
+                // self.cfg.actor.micro_batch_size
+                // self._world_size
+            )
+
+            metrics = {}
+
+            for _ in range(self.gradient_accumulation):
+                observation, actions = next(self.data_iter)
```
Comment (guozhen1997) [inline rlinf/workers/sft/fsdp_sft_worker.py:116]: The distributed dataloader already yields batches of size `global_batch_size // world_size`, so accumulating gradients in this loop is incorrect.

File: examples/sft/run_embodiment_sft.sh (hunk 13 above)
Comment (XuS1994) [reply examples/sft/run_embodiment_sft.sh]: Done

File: rlinf/models/embodiment/openpi/policies/maniskill_policy.py (hunk 14 above)
Comment (XuS1994) [reply rlinf/models/embodiment/openpi/policies/maniskill_policy.py]: Done
